### PR TITLE
feat(validation): domain ③ Phase 1 — data-prep checks (boundary_leakage + prep_published_parity)

### PR DESCRIPTION
**Pull request by saiemgilani**
## Summary

Phase 1 (data-side) of **domain ③ — data-prep validation**, the next domain in the workspace validation program. **Stacked on the validation harness (#141)** — base branch is `feat/validation-harness-tier1`, so this PR shows only the four data-prep commits and should merge after #141.

Adds two language-agnostic data-prep checks to the Tier-1 engine, both emitting the existing `Finding` contract.

## What's included

- **`boundary_leakage`** (`checks/boundary_leakage.py`) — single-frame check (joins the CLI `run` loop). Flags **lag columns that are non-null on a group's first row** — a cross-game leak (e.g. a `prev_ep` carried from the prior game). Detected via `pl.int_range(pl.len()).over(group_key) == 0` (frame assumed play-ordered within each group).
- **`prep_published_parity`** (`checks/prep_published_parity.py`) — two-frame check (harness-driven, like `e2e`; **not** in the single-frame `_CHECKS`). Flags prep key groups dropped from published (anti-join) and per-numeric-column value divergence beyond tolerance — including the **null-side** case (a value dropped to null on one side).
- **Registry fields** — `lag_columns`, `cumulative_columns`, `group_key` on both `DatasetSpec` and `CheckContext`, threaded through `_resolve_spec`.
- Both new modules on the `[tool.mypy] files` ratchet.

## Gate

`uv run mypy` → Success (28 files) · `uv run pytest tests/contracts/` → 38 passing · ruff check + format clean.

## Review note

The whole-branch review caught two silent-failure bugs in `prep_published_parity` (both now fixed in `9b93820`): a **null-divergence miss** (a prep value dropped to `null` in published reported as parity) and a **suffix-collision crash** (`DuplicateError` if `published` ships a `<col>_published` column → now a collision-proof `__pub__` suffix). Both aligned with the harness's own "fail toward surfacing" charter.

## Deferred (by design — not gaps)

- Cumulative-column boundary heuristic (Phase 1.5; its reset semantics are column-specific). `cumulative_columns` is accepted on the spec but not yet inspected.
- `nfl_model_pbp` dataset registration with real lag/cumulative columns (`boundary_leakage` is fully fixture-tested without it).
- Source linters (Python `ast` / R `Rscript`) — Phases 2–3, separate PRs.

## Notes

- Design spec + plan live in `dev/` (gitignored internal notes, per repo convention).

## Summary by Sourcery

Introduce data-prep validation checks for boundary leakage and prep/published parity and thread required leakage metadata through the validation registry and context.

New Features:
- Add boundary_leakage single-frame check to flag non-null lag values on group boundary rows.
- Add prep_published_parity two-frame check to detect dropped prep key groups and numeric value divergence between prep and published frames.

Enhancements:
- Extend DatasetSpec and CheckContext with lag_columns, cumulative_columns, and group_key fields to support leakage-aware validation.
- Register the new data-prep checks with the CLI tier-1 engine and mypy ratchet configuration.

Tests:
- Add contract tests covering boundary_leakage behavior across leak, clean, and skip scenarios.
- Add contract tests ensuring prep_published_parity detects dropped rows, numeric divergence, and null-side divergence.

### CODE REVIEW

File: tests/contracts/test_prep_published_parity.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+def test_matching_frames_yield_no_findings():
+    prep = pl.DataFrame({"game_id": [1, 2], "epa": [0.1, 0.2]})
+    published = pl.DataFrame({"game_id": [1, 2], "epa": [0.1, 0.2]})
+    assert prep_published_parity.run("d", prep, published, ("game_id",), "nfl") == []
+
+
+def test_dropped_row_is_error():
+    prep = pl.DataFrame({"game_id": [1, 2], "epa": [0.1, 0.2]})
+    published = pl.DataFrame({"game_id": [1], "epa": [0.1]})  # game 2 dropped
+    findings = prep_published_parity.run("d", prep, published, ("game_id",), "nfl")
```
[the remark this answers is not in the document]
Comment (saiemgilani) [reply tests/contracts/test_prep_published_parity.py:16]: Fixed in 05d2e935: added `test_within_tolerance_yields_no_findings` to `tests/contracts/test_prep_published_parity.py` — diffs below `tolerance=1e-6` assert zero findings, pinning the suppression path.

[the remark this answers is not in the document]
Comment (saiemgilani) [reply tests/contracts/test_prep_published_parity.py:16]: Fixed in 05d2e935: (1) `prep_published_parity.run()` now renames published columns that already end with `"__pub__"` before the join, preventing `DuplicateError`; (2) added `test_suffix_collision_still_reports_divergence` which asserts divergence in `"epa"` is still detected when published carries `"epa__pub__"` — pinning the regression.

File: tests/contracts/test_boundary_leakage.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+def test_lag_nonnull_on_first_of_game_is_error():
+    # game 1 first row has a carried prev_ep (leak); proper resets are null on row 0
+    frame = pl.DataFrame(
+        {
+            "game_id": [1, 1, 2, 2],
+            "prev_ep": [3.5, 1.0, None, 2.0],  # game 1 row0 = 3.5 (LEAK); game 2 row0 = None (ok)
+        }
+    )
+    findings = boundary_leakage.run("nfl_pbp", frame, _ctx())
+    assert any(f.severity is Severity.ERROR and "prev_ep" in f.message for f in findings)
```
[the remark this answers is not in the document]
Comment (saiemgilani) [reply tests/contracts/test_boundary_leakage.py:22]: Applied in 05d2e935: added `test_two_lag_columns_both_detected` — both `prev_ep` and `prev_wp` leaks at group boundaries are independently asserted.

File: tests/contracts/test_boundary_leakage.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+
+def test_clean_lag_resets_yield_no_findings():
+    frame = pl.DataFrame(
+        {
+            "game_id": [1, 1, 2, 2],
+            "prev_ep": [None, 1.0, None, 2.0],  # both games reset to null on row 0
+        }
+    )
+    assert boundary_leakage.run("nfl_pbp", frame, _ctx()) == []
```
[the remark this answers is not in the document]
Comment (saiemgilani) [reply tests/contracts/test_boundary_leakage.py:32]: Applied in 05d2e935: added `test_empty_lag_columns_returns_no_findings` — verifies the early-exit path when `lag_columns=()` is configured, even if the frame has a non-null lag column present.

**Comment by saiemgilani** [comment]
Superseded by #149, which re-linearizes this stack's work (with all bot-review fixes) onto main — the divergent stacked-squash dropped content, so the chain was consolidated. Closing as landed via #149.
